### PR TITLE
implemented scale action (#615)

### LINKED ISSUE #615: As a user I want to be able to scale a deployment up or down without editing the yaml
https://plugins.jetbrains.com/plugin/10485-kubernetes/reviews#review=80422
image

### PR DESCRIPTION
**Pull request by adietish**
fixes #615

### CODE REVIEW

**Comment by adietish** [comment]
To test this, one needs a Kubernetes and an OpenShift cluster. One should create the following resources: 
* Deployment
```yaml
apiVersion: "apps/v1"
kind: "Deployment"
metadata:
  labels:
    app: "nginx"
    smurf: "grumpy"
    jedi: yoda
  name: "nginx2"
  namespace: "default"
spec:
  replicas: 3
  selector:
    matchLabels:
      app: "nginx"
  template:
    metadata:
      labels:
        app: "nginx"
    spec:
      containers:
      - image: "nginx"
        name: "nginx1"
        ports:
        - containerPort: 80
      - image: "nginx"
        name: "nginx2"
        ports:
        - containerPort: 8080
```
* DeploymentConfig (only OpenShift)
```yaml
apiVersion: apps.openshift.io/v1
kind: DeploymentConfig
metadata:
  annotations:
    app.openshift.io/vcs-url: <<unknown>>
    app.quarkus.io/build-timestamp: 2023-07-19 - 12:59:17 +0000
  labels:
    app.kubernetes.io/name: openshift-quickstart
    app.kubernetes.io/version: 1.0.0-SNAPSHOT
    app.kubernetes.io/managed-by: quarkus
    app.openshift.io/runtime: quarkus
  name: openshift-quickstart
spec:
  replicas: 2
  selector:
    app.kubernetes.io/version: 1.0.0-SNAPSHOT
    app.kubernetes.io/name: openshift-quickstart
  template:
    metadata:
      annotations:
        app.openshift.io/vcs-url: <<unknown>>
        app.quarkus.io/build-timestamp: 2023-07-19 - 12:59:17 +0000
      labels:
        app.kubernetes.io/managed-by: quarkus
        app.openshift.io/runtime: quarkus
        app.kubernetes.io/version: 1.0.0-SNAPSHOT
        app.kubernetes.io/name: openshift-quickstart
    spec:
      containers:
        - env:
            - name: QUARKUS_LOG_LEVEL
              value: ALL
            - name: KUBERNETES_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
          image: openshift-quickstart:1.0.0-SNAPSHOT
          imagePullPolicy: Always
          name: openshift-quickstart
          ports:
            - containerPort: 8080
              name: http
              protocol: TCP
            - containerPort: 8443
              name: https
              protocol: TCP
  triggers:
    - imageChangeParams:
        automatic: true
        containerNames:
          - openshift-quickstart
        from:
          kind: ImageStreamTag
          name: openshift-quickstart:1.0.0-SNAPSHOT
      type: ImageChange
```
* ReplicationController
```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx
spec:
  replicas: 2
  selector:
    app: nginx-rc
  template:
    metadata:
      name: nginx-rc
      labels:
        app: nginx-rc
    spec:
      containers:
      - name: nginx-rc
        image: nginx
        ports:
        - containerPort: 80
```
* ReplicaSet
```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: frontend
  labels:
    app: guestbook
    tier: frontend
spec:
  replicas: 3
  selector:
    matchLabels:
      tier: frontend
  template:
    metadata:
      labels:
        tier: frontend
    spec:
      containers:
      - name: php-redis
        image: gcr.io/google_samples/gb-frontend:v3
```
Once created one can then select any of the following resoures and choose "Scale" in the context menu:
* Deployment
* DeploymentConfig (only OpenShift)
* ReplicationController
* ReplicaSet
* element is Pod

**Comment by adietish** [comment]
> It would be nice to prevent scaling a `ReplicationController` that is a child of a `DeploymentConfig`.

Awwesome idea. Will try this!

**Comment by adietish** [comment]
@datho7561 When selecting a `ReplicationController` which has a parent `DeploymentConfig` I scale the `DeploymentConfig`. If there's non, the `ReplicationController` is scaled. The dialog now tells you what resource is being scaled. 
Can you please verify again?

**Comment by adietish** [comment]
> Okay, it scales the DeploymentConfig now, but unfortunately, the tree sometimes isn't updated with the new Pods that are created.

This happened to me once and then I couldn't reproduce it. Trying to see what's wrong.

**Comment by adietish** [comment]
@datho7561 There's one item I'd love to get your feedback on: In the Scale replica dialog I print the scaled resource in the 1st line of the dialog:
image
I did not add anything like "Resource: " in front of the label or specify the operation (ex. "Scale DeploymentConfig openshift-quickstart"). I'm trying to keep texts short, concise. That this dialog is about scaling is already visible in the dialog title. Then the kind is printed in front of the name which is - I think - sufficient to grasp that this is the resource that this concerns. Would you agree?

**Comment by datho7561** [comment]
I think it's fine without "Resource" in front.

**Comment by adietish** [comment]
@datho7561 can you please re-verify if you still face the lacking refresh?

**Review by datho7561 — APPROVED** [review]
It seems to be working well now! I haven't had a chance to look over the code, let me know if you would like me to do that.

**Comment by adietish** [comment]
@datho7561 sure, if you can review the code, that would be awesome!

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/kubernetes/KubernetesReplicas.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import com.redhat.devtools.intellij.kubernetes.model.resource.openshift.ReplicationControllersOperator
+import io.fabric8.kubernetes.api.model.HasMetadata
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.api.model.ReplicationController
+import io.fabric8.kubernetes.api.model.apps.Deployment
+import io.fabric8.kubernetes.api.model.apps.ReplicaSet
+import io.fabric8.kubernetes.api.model.apps.StatefulSet
+
+open class KubernetesReplicas(
+    protected val operator: NonCachingSingleResourceOperator,
```
Comment (datho7561) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/kubernetes/KubernetesReplicas.kt:25]: `operator` appears to be a reserved word in Kotlin. It doesn't appear to be causing any issues in this case, so I don't think changing it is necessary.

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/openshift/OpenShiftReplicas.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import com.redhat.devtools.intellij.kubernetes.model.context.IActiveContext
+import com.redhat.devtools.intellij.kubernetes.model.resource.NonCachingSingleResourceOperator
+import com.redhat.devtools.intellij.kubernetes.model.resource.kubernetes.KubernetesReplicas
+import io.fabric8.kubernetes.api.model.HasMetadata
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.api.model.ReplicationController
+import io.fabric8.openshift.api.model.DeploymentConfig
+
+class OpenShiftReplicas(operator: NonCachingSingleResourceOperator, getAllResources: ResourcesRetrieval)
```
Comment (datho7561) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/openshift/OpenShiftReplicas.kt:21]: Another instance of `operator` as a variable name

**Review by datho7561 — APPROVED** [review]
The code looks pretty good! Found something to comment about it, but I think the code is good as is.

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/openshift/OpenShiftReplicas.kt (hunk 2 above)
Comment (adietish) [reply src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/openshift/OpenShiftReplicas.kt]: corrected

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/kubernetes/KubernetesReplicas.kt (hunk 1 above)
Comment (adietish) [reply src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/resource/kubernetes/KubernetesReplicas.kt]: corrected

**Pull request merged**